### DISCUSSION TITLE
ensure_leading_slash and virtual hosting

### DISCUSSION THREAD

**Discussion by yuzuquats** [actix/actix-web#3278] [Q&A] [answered]
I'm following the [example](https://actix.rs/docs/application/#application-guards-and-virtual-hosting) in the actix.rs website where  `web::scope("/")` is used. Unfortunately, it seems as though adding a second route is non-trivial because `Resource` automatically prepends a leading slash:
```rust
web::scope("/")
    .guard(guard::Host("www.rust-lang.org"))
    .route("", web::to(|| async { HttpResponse::Ok().body("www") }))
    .route("hello", web::to(|| async { HttpResponse::Ok().body("www") }))       // <--- adding this routes to //hello
    // .route("/hello", web::to(|| async { HttpResponse::Ok().body("www") }))   // <--- alternatively, adding this *also* routes to //hello
```

This means that in the above example, `/` and `//hello` will route. `/` and `/hello` will not, however.  Important to note, adding the `actix_web::middleware::NormalizePath::trim()` middleware means `hello` will be impossible to route to afaik.

I tried a couple of solutions, but i can't think of a great way to create a collection of routes with a shared scope + guard. Couple of options:

1) Guard every route individually
```rust
.service(
  web::scope("/").guard(Host("host1.com")).route(...)
)
.service(
  web::scope("/").guard(Host("myhost.com")).route(...)
)
.service(
  web::scope("/hello").guard(Host("host1.com")).route(...)
)
```
2) guard *everything* (ie. "") instead of "/"
```rust
.service(
  web::scope("").guard(Host("host1.com")).route("/", ...)
)
.service(
  web::scope("").guard(Host("myhost.com")).route("/", ...).route("/hello", ...)
)
```
3) add a custom middleware that specifically routes "/hello" to "//hello" (like NormalizePath but in reverse?)

None of these options are optimal, so I'm not sure if I'm just missing something obvious. Is there a reason why Resources prepends leading slashes and is it possible to override that behavior?

**Comment by robjtede** [comment, marked as answer]
The semantics of resource definitions is defined in detail over here: <https://docs.rs/actix-web/4.5.1/actix_web/dev/struct.ResourceDef.html#pattern-format-and-matching-behavior>

In particular:

> This means that pattern string must either be empty or begin with a slash (/).

Therefore this behavior is intentional. The canonical way to avoid double slashes here is to use `web::scope("")`. This is also documented: <https://docs.rs/actix-web/4.5.1/actix_web/struct.Scope.html#avoid-trailing-slashes>.

> Avoid using trailing slashes in the scope prefix (e.g., `web::scope("/scope/")`). It will almost certainly not have the expected behavior.
